### PR TITLE
Add Bitbucket support (text prompts only)

### PR DESCRIPTION
**Pull request by mjcheetham**
Add a host provider to support username/password and OAuth authentication to Atlassian Bitbucket.
The auth flow is based on the logic from GCM Windows, and follows the following chart:

```
            Yes   +---------+   No
     +------------+ Has RT? +-------------+
     |            +---------+             |
     v                                    v
+----+----+                        +------+------+
|  OAuth  |   Fail                 |     Has     | No
| Refresh +---------+              | credential? +----+
+---+-----+         |              +--+----------+    |
    |               |                 |               |
OK  |               |             Yes |               |
    v               v                 v               v
+---+---+       +---+----+   Yes   +--+---+    +------+-+
| Store +<------+ OAuth  +<--------+ 2FA? +<---+  U/P   |
|  RT   |       | Prompt |         +--+---+    | Prompt |
+---+---+       +--------+            |        +--------+
    |                                 | No
    v                                 v
+---+----+                         +--+---------+
| Return |                         |   Return   |
|   AT   |                         | credential |
+--------+                         +------------+

```

- We first check for the presence of an OAuth refresh token, and if it exists we use it to create a fresh access token.
- If we fail to use the refresh token then we restart the OAuth flow.
- If we do not have a refresh token then we try and use any existing credential and check if 2FA is required.
- If 2FA is required we start an OAuth flow.

There are no GUI prompts in this PR; text-based prompts are used instead. A port of the Windows Bitbucket UI will follow this PR.

Goes towards implementing #37

### CODE REVIEW

File: src/shared/Atlassian.Bitbucket/BitbucketConstants.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace Atlassian.Bitbucket
+{
+    public static class BitbucketConstants
+    {
+        public const string BitbucketBaseUrlHost = "bitbucket.org";
+        public static readonly Uri BitbucketApiUri = new Uri("https://api.bitbucket.org");
+        public const string BitbucketAuthHelperName = "Atlassian.Bitbucket.UI";
+
+        public const string OAuth2ClientId = "b5AKdPfpgFdEGpKzPE";
```
Comment (mjcheetham) [inline src/shared/Atlassian.Bitbucket/BitbucketConstants.cs:13]: This is a new application "Git Credential Manager" still 'owned' by myself. We can swap this out for the app ID used by the old GCM, which was the same as Sourcetree's! I'd like to try and reach out and whitelist our application (and make it more Microsoft 'official').

Comment (derrickstolee) [reply src/shared/Atlassian.Bitbucket/BitbucketConstants.cs:13]: and this is registered through BitBucket?

Comment (mjcheetham) [reply src/shared/Atlassian.Bitbucket/BitbucketConstants.cs:13]: Yes. This is registered through Bitbucket. I've started the process of transferring the GitHub GCM app to a verified organisation (following MS policy), so I'd need to do the same for Bitbucket, but that's new ground AFAIK. Perhaps the Azure Pipelines folks know more here?
The alternative is to continue to 'steal' Sourcetree's ID?

Comment (mjcheetham) [reply src/shared/Atlassian.Bitbucket/BitbucketConstants.cs]: I'll continue to use Sourcetree's values until we can do this.

**Comment by mminns** [comment]
I wrote the original GCMW implementation, I'd be happy to help test this.

**Comment by mjcheetham** [comment]
> Are there any existing build artefacts are should I just build from master?

Hi @mminns! 
You'll probably want to build from a branch in my fork (there is a bug where I forgot to include the actual Bitbucket host provider DLL needed! 🤦) until this PR is completed: #117 

https://github.com/mjcheetham/Git-Credential-Manager-Core/tree/fix-bb

## Windows
1. Build the Installer.Windows project in the WindowsDebug configuration inside VS
2. Look in `out/windows/Installer.Windows/bin/Debug` and you'll find the exe installer

## Mac
1. Run `dotnet build -c MacDebug` from the root of the repo.
2. Look in `out/osx/Installer.Mac/pkg/Debug` and you'll find the pkg installer

This build only has a command-line auth prompt experience and not the WPF-based UI. This is coming in this PR: #116.

**Pull request merged**